### PR TITLE
Dev environments

### PR DESCRIPTION
**Pull request by eguzki**
### What

Docker compose based development environment. Looking for easy testing setups for different use cases.

Environments:
* Gateway listening on TLS: path `dev-environments/listen-tls` 
* PROXY with upstream using TLSv1.3: path `dev-environments/https-proxy-upstream-tlsv1.3`
* PROXY with upstream using plain HTTP 1.1: path `dev-environments/http-proxy-plain-http-upstream`
* Gateway instrumented with opentelemetry: path `dev-environments/opentelemetry-instrumented-gateway`
* Plain HTTP 1.1 upstream: path `dev-environments/plain-http-upstream`
* [Upstream using TLSv1.3: path `dev-environments/upstream-tlsv1.3`

Usually dev environment can run by`cd`-ing to the path and running `make gateway`. For example:

```
cd dev-environments/listen-tls
# generate certs
make certs
make gateway IMAGE_NAME=quay.io/3scale/apicast:latest
```

If the environment requires TLS certificates, there is an extra step (Makefile target) to generate on the fly the (self signed) certificates. I have tried to avoid committing certs to github, even if they are fake for `example.com`.

### CODE REVIEW

File: dev-environments/https-proxy-upstream-tlsv1.3/tinyproxy.Dockerfile
Comment (tkan145) [inline dev-environments/https-proxy-upstream-tlsv1.3/tinyproxy.Dockerfile:1]: I was not able to build tiny proxy from this Dockerfile

```
[+] Building 2.4s (5/5) FINISHED                                                                                                    
 => [proxy internal] load build definition from tinyproxy.Dockerfile                                                                
 => => transferring dockerfile: 627B                                                                                                
 => [proxy internal] load .dockerignore                                                                                             
 => => transferring context: 2B                                                                                                     
 => [proxy internal] load metadata for docker.io/library/alpine:3.17.3                                                              
 => CACHED [proxy 1/2] FROM docker.io/library/alpine:3.17.3@sha256:124c7d2707904eea7431fffe91522a01e5a861a624ee31d03372cc1d138a3126 
 => ERROR [proxy 2/2] RUN apk --no-cache add tinyproxy=1.11.1-r2                                                                    
------                                                                                                                              
 > [proxy 2/2] RUN apk --no-cache add tinyproxy=1.11.1-r2:                                                                          
0.248 fetch https://dl-cdn.alpinelinux.org/alpine/v3.17/main/x86_64/APKINDEX.tar.gz                                                 
0.249 fetch https://dl-cdn.alpinelinux.org/alpine/v3.17/community/x86_64/APKINDEX.tar.gz                                            
0.249 WARNING: Ignoring https://dl-cdn.alpinelinux.org/alpine/v3.17/main: IO ERROR                                                  
0.250 WARNING: Ignoring https://dl-cdn.alpinelinux.org/alpine/v3.17/community: IO ERROR                                             
0.250 ERROR: unable to select packages:                                                                                             
0.251   tinyproxy (no such package):                                                                                                
0.251     required by: world[tinyproxy=1.11.1-r2]                                                                                   
------                                                                                                                              
failed to solve: process "/bin/sh -c apk --no-cache add tinyproxy=1.11.1-r2" did not complete successfully: exit code: 1            
make: *** [Makefile:9: gateway] Error 17                                                                                            

```

**Review by tkan145 — COMMENTED** [review]
The rest look good to me

I also want to ask what will be added to this folder? I can imagine that this directory would grow very quickly if we added a new folder for each use case.

File: dev-environments/https-proxy-upstream-tlsv1.3/tinyproxy.Dockerfile
Comment (eguzki) [reply dev-environments/https-proxy-upstream-tlsv1.3/tinyproxy.Dockerfile:1]: I have cleaned my docker env, all images and containers. I could successfully run the environment

```
=> [proxy internal] load build definition from tinyproxy.Dockerfile                                                                                                                                                                                                          0.0s
 => => transferring dockerfile: 551B                                                                                                                                                                                                                                          0.0s
 => [proxy internal] load .dockerignore                                                                                                                                                                                                                                       0.0s
 => => transferring context: 2B                                                                                                                                                                                                                                               0.0s
 => [proxy internal] load metadata for docker.io/library/alpine:3.17.3                                                                                                                                                                                                        1.4s
 => [proxy 1/2] FROM docker.io/library/alpine:3.17.3@sha256:124c7d2707904eea7431fffe91522a01e5a861a624ee31d03372cc1d138a3126                                                                                                                                                  0.5s
 => => resolve docker.io/library/alpine:3.17.3@sha256:124c7d2707904eea7431fffe91522a01e5a861a624ee31d03372cc1d138a3126                                                                                                                                                        0.0s
 => => sha256:124c7d2707904eea7431fffe91522a01e5a861a624ee31d03372cc1d138a3126 1.64kB / 1.64kB                                                                                                                                                                                0.0s
 => => sha256:b6ca290b6b4cdcca5b3db3ffa338ee0285c11744b4a6abaa9627746ee3291d8d 528B / 528B                                                                                                                                                                                    0.0s
 => => sha256:9ed4aefc74f6792b5a804d1d146fe4b4a2299147b0f50eaf2b08435d7b38c27e 1.47kB / 1.47kB                                                                                                                                                                                0.0s
 => => sha256:f56be85fc22e46face30e2c3de3f7fe7c15f8fd7c4e5add29d7f64b87abdaa09 3.37MB / 3.37MB                                                                                                                                                                                0.4s
 => => extracting sha256:f56be85fc22e46face30e2c3de3f7fe7c15f8fd7c4e5add29d7f64b87abdaa09                                                                                                                                                                                     0.1s
 => [proxy 2/2] RUN apk --no-cache add tinyproxy=1.11.1-r2                                                                                                                                                                                                                    0.8s
 => [proxy] exporting to image                                                                                                                                                                                                                                                0.0s 
 => => exporting layers                                                                                                                                                                                                                                                       0.0s 
 => => writing image sha256:4a2c5d27b4a9a8bb3d1586baa1176772a8bbb324d44faf9b876ce7e691ad9bfa                                                                                                                                                                                  0.0s 
 => => naming to docker.io/library/https-proxy-upstream-tlsv13-proxy
```

Maybe temporary error?? (can you try again?) Maybe arch issue? I am running on x86_64 host

Comment (tkan145) [reply dev-environments/https-proxy-upstream-tlsv1.3/tinyproxy.Dockerfile:1]: turns out it was a problem with my RHEL machine,  ran fine with my Fedora box.
 
+1

**Comment by tkan145** [comment]
Initially, I thought we could share common files, but having independent files also has its own advantages. I think it's good as is and if it grows we can always change or move somewhere else later

**Comment by tkan145** [comment]
Look good to me as is, I think we can merge this and improve later if we need to

**Review by tkan145 — APPROVED** [review]

**Pull request merged**